### PR TITLE
Notebook#3: Adapter for ensembling

### PR DESCRIPTION
**Pull request by grzes314**

### CODE REVIEW

File: examples/3-adapter_advanced.ipynb [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+    "The basic usage is as follows:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "rename_step = Step(name='rename',\n",
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: Maybe `renamed_step`?

Comment (mromaniukcdl) [reply examples/3-adapter_advanced.ipynb]: (instead of `rename_step`)

Comment (mromaniukcdl) [reply examples/3-adapter_advanced.ipynb]: Never mind, I didn't notice that this step is a dummy renaming step. Please ignore.

File: examples/3-adapter_advanced.ipynb [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+   "outputs": [],
+   "source": [
+    "xgb_step.fit_transform(data_train)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "All right, so now we have two models trained, but we haven't checked so far how do they perform. Let's do it now. We will use one of data scientists' favourite measures: log-loss."
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: Do we really have to use log-loss? It seems to complicate things and distract from the main point.

File: examples/3-adapter_advanced.ipynb [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "guesses_step = Step(name='guesses_maker',\n",
+    "                 transformer=Dummy(),\n",
+    "                 input_steps=[ensemble_step],                 \n",
+    "                 adapter={'y_pred': ([(ensemble_step.name, 'y_proba')],\n",
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: Are we making a step where the adapter does all the computation and then the transformer does nothing? If it's easier to do it this way then we may have to think about re-designing the API...

File: examples/3-adapter_advanced.ipynb [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ensemble_step = Step(name='ensembler',\n",
+    "                 transformer=Dummy(),\n",
+    "                 input_steps=[rf_step, xgb_step],                 \n",
+    "                 adapter={'y_proba': ([(rf_step.name, 'y_proba'),\n",
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: So we made the adapter do all the computation and the transformer does nothing? This seems counter-intuitive to me...

File: examples/3-adapter_advanced.ipynb [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+    "In one of the previous cells we checked quality of our model manually. Let's add a final step that will do it for us automatically!"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class EvaluationTransformer(BaseTransformer):\n",
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: Hmm... making a step that does evaluation is very clever but It makes me really confused. @jakubczakon and @kamil-kaczmarek, what do you think?

File: examples/3-adapter_advanced.ipynb [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+   "metadata": {},
+   "source": [
+    "### Peek on pipeline predictions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "It is always very pleasant to compare images with model's predictions. As a last example we show a step that displays a few images with the predicted probability distributions!"
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: Maybe just 'nice' instead of 'pleasant'?

File: examples/3-adapter_advanced.ipynb [hunk 7]
```diff
@@ -0,0 +1,740 @@
...
+   "outputs": [],
+   "source": [
+    "evaluation_step.transform(data_test)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Peek on pipeline predictions"
```
Comment (mromaniukcdl) [inline examples/3-adapter_advanced.ipynb]: 'Peek' sounds weird. I would much prefer 'Looking at pipeline predictions' or maybe 'viewing ...' or 'inspecting ...'

File: examples/3-adapter_advanced.ipynb (hunk 5 above)
Comment (jakubczakon) [reply examples/3-adapter_advanced.ipynb]: @mromaniukcdl @kamil-kaczmarek @grzes314 We were discussing it with @kamil-kaczmarek a while back. I have to say that I like to have pipeline (model) and evaluation as two separate concepts. It bears less risk of data leakage, You can build various pipelines and evaluate them with the same evaluation scheme. However using steps to do everything from fetching data from db and setting up the CV scheme all the way through producing reports and sending an email to the manager is not out of the picture here. I would distinguish however, pipelines from workflows where the first is pure modeling, second however does everything. Notice that pipeline can easily be just a subset of the workflow.

File: examples/3-adapter_advanced.ipynb [hunk 8]
```diff
@@ -0,0 +1,740 @@
...
+    "The next step incorporates averaging of model predictions into the pipeline."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ensemble_step = Step(name='ensembler',\n",
```
Comment (jakubczakon) [inline examples/3-adapter_advanced.ipynb]: @grzes314 I have to say that I usually have this as a trainable step (xgboost or something) or prediction average that is done in .transform. If we were to go with the adapter+Dummy (no operation) usage on a regular basis I would push for a separate concept: Adapter class .

**Comment by jakubczakon** [comment]
I have to say that I really like that you guys are stretching and poking around this project. It is really eye openning. 
Also @mromaniukcdl I am stuck in between the "API is confusing (and bad)" since you used it differently and "steps can be way more than sklearn pipelines"

**Comment by grzes314** [comment]
Following the code review and our discussion during stand-up, I made quite a few changes in the notebook:
- Instead of creating RandomForest and XGBoost which are ensembled by simple averaging, I create a few RandomForests and ensemble them with XGBoost.
- Because ensembling is now supervised, I split the data into three parts: for training, ensembling and testing.
- I got rid of `Dummy` transformers.
- I got rid of `force_fitting`
